specifications/cosmwasm/SPEC.md

```diff
@@ -0,0 +1,1413 @@
+        value: Binary,
+    },
+    Ibc(IbcMsg),
+    Wasm(WasmMsg),
+} 
+```
+
+Actors define an entrypoint for dealing with messages and decide how to process them and alter its internal state.
+
+```rust
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> { }
+```
+### 4.2.1 Submessages
+Messages are used to interact with modules or other smart contracts.
+Given the actor model and the loose coupling between actors, messages are sent but no state is kept.
+This means there is no response on whether if the call result was successful or not.
+
+In some scenarios, the ability to get a result and decide whether the transaction MUST be aborted or not is extremely useful.
+For this purpose, submessages are introduced: a second message sent with the sole purpose of getting a call result from a previous message.
+
+```rust
+pub struct SubMsg<T> {
+    pub id: u64,                // reply_id that will be used to handle the reply
+    pub msg: CosmosMsg<T>,      // message to be sent
+    pub gas_limit: Option<u64>, // gas limit for the submessage
+    pub reply_on: ReplyOn,      // a flag to determine when the reply should be sent
+}
+```
+Submessages offer different patterns depending on the strategy to handle replies.
+Please note that a sub-message failure always induces a revert of the sub-transaction that executed it.
+Therefore, you get the result but the state on the destination actor has been already committed or reverted.
+This does not revert the calling contract state.
+It will simply return an error the calling contract will decide how to handle according to the following patterns:
+```rust
+pub enum ReplyOn {
+    /// Always perform a callback after SubMsg is processed
+    Always,
+    /// Only callback if SubMsg returned an error, no callback on success case
+    Error,
+    /// Only callback if SubMsg was successful, no callback on error case
+    Success,
+    /// Never make a callback - this is like the original CosmosMsg semantics
+    Never,
+}
+```
+If the caller decides to not handle the error, its state will be reverted. 
+
+### 4.2.2 Query
+Queries are read-only messages.
+They are used to query the state of another actor in the system.
+
+```rust
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum QueryMsg {
+  // ResolveAddress returns the current address that the name resolves to
+  ResolveRecord { name: String },
+  Config {},
+} 
+```
+
+## 4.3. Calling a Contract
+Contract to contract calls are structured through messaging.
+Calling another contract requires messages to go through a `Dispatcher` that maintains the state related to the calls.
+The purpose of the dispatcher is twofold: first, it prevents contracts from calling other contracts directly to prevent reentrancy attacks.
+Second , it helps to maintain atomicity and revert if a call to another contract fails.
```

Comment:
It could be added. However, we tried to limit the list of terms and definitions to Cosmwasm only terms.
We assumed that general blockchain terms would be already known by the potential reader. Unless atomicity has a special connotation in CosmWasm, I would not add it.